src/args.rs

```diff
@@ -0,0 +1,441 @@
+///
+/// Note that the field names are relevant for the TOML's tables.
+///
+/// TODO: Figure out how to adapt file isolation.
+#[derive(Debug, Deserialize)]
+pub struct UhyveGuestConfig {
+	pub memory: MemoryArgs,
+	pub cpu: CpuArgs,
+	pub guest: GuestArgs,
+}
+
+use merge::Merge;
+
+#[derive(Default, Merge, Parser, Debug, Deserialize)]
+pub struct MemoryArgs {
+	/// Guest RAM size
+	#[clap(short = 'm', long, env = "HERMIT_MEMORY_SIZE")]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub memory_size: Option<GuestMemorySize>,
+
+	/// Disable ASLR
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub no_aslr: Option<bool>,
+
+	/// Transparent Hugepages
+	///
+	/// Advise the kernel to enable Transparent Hugepages [THP] on the virtual RAM.
+	///
+	/// [THP]: https://www.kernel.org/doc/html/latest/admin-guide/mm/transhuge.html
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub thp: Option<bool>,
+
+	/// Kernel Samepage Merging
+	///
+	/// Advise the kernel to enable Kernel Samepage Merging [KSM] on the virtual RAM.
+	///
+	/// [KSM]: https://www.kernel.org/doc/html/latest/admin-guide/mm/ksm.html
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub ksm: Option<bool>,
+}
+
+/// Arguments for the CPU resources allocated to the guest.
+#[derive(Default, Debug, Clone, Merge, Parser, Deserialize)]
+pub struct CpuArgs {
+	/// Number of guest CPUs
+	#[clap(short, long, env = "HERMIT_CPU_COUNT")]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub cpu_count: Option<CpuCount>,
+
+	/// Create a PIT
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub pit: Option<bool>,
+
+	/// Bind guest vCPUs to host cpus
+	///
+	/// A list of host CPU numbers onto which the guest vCPUs should be bound to obtain performance benefits.
+	/// List items may be single numbers or inclusive ranges.
+	/// List items may be separated with commas or spaces.
+	///
+	/// # Examples
+	///
+	/// * `--affinity "0 1 2"`
+	///
+	/// * `--affinity 0-1,2`
+	#[clap(short, long, name = "CPUs")]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub affinity: Option<Affinity>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Affinity(Vec<CoreId>);
+
+impl Affinity {
+	pub fn parse_ranges_iter<'a>(
+		ranges: impl IntoIterator<Item = &'a str> + 'a,
+	) -> impl Iterator<Item = Result<usize, ParseIntError>> + 'a {
+		struct ParsedRange(RangeInclusive<usize>);
+
+		impl FromStr for ParsedRange {
+			type Err = ParseIntError;
+
+			fn from_str(s: &str) -> Result<Self, Self::Err> {
+				let range = match s.split_once('-') {
+					Some((start, end)) => start.parse()?..=end.parse()?,
+					None => {
+						let idx = s.parse()?;
+						idx..=idx
+					}
+				};
+				Ok(Self(range))
+			}
+		}
+
+		ranges
+			.into_iter()
+			.map(ParsedRange::from_str)
```

Comment:
Defining a new type, only to use it once and not even using some derived trait behavior looks a bit overengineered.

Reply:
See also #1036.
 
 (this thread is resolved)